transport_plugins/jlink/iotile_transport_jlink/jlink_background.py

```diff
@@ -358,19 +384,92 @@ def _read_memory_blocking(self, start_address, length, chunk_size=4, join=True):
+        if not self._jlink_adapter.opened_interfaces["rpc"]:
+            return
+
         poll_address, poll_mask = control_info.poll_info()
         value, = await self.read_memory(poll_address, 1, chunk_size=1)
 
         if value & poll_mask:
             self.rpc_response_event.set()
 
+    async def _read_next_queue_frame(self, control_info, read_index):
+        header_address, data_address = control_info.queue_element_info(read_index)
+        header, = await self.read_memory(header_address, 1, chunk_size=1)
+        data = await self.read_memory(data_address, control_info.FRAME_SIZE - 1, chunk_size=1)
+
+        is_trace = header & 0x40
+        is_stream = header & 0x80
+        data_length = header & 0x3F
+
+        if data_length > ControlStructure.FRAME_SIZE - 1:
+            raise HardwareError("Data length is too big {}".format(data_length))
+
+        if is_trace and self._jlink_adapter.opened_interfaces["tracing"]:
+            self._jlink_adapter.add_trace(data[0:data_length])
+        elif is_stream and self._jlink_adapter.opened_interfaces["streaming"]:
+            self._jlink_adapter.report_parser.add_data(data[0:data_length])
+        else:
+            pass # Drop if interface is not opened
+
+    async def _poll_queue_status(self, control_info):
+        """ Read next fream from queue
+
+            Returns:
+                bool: true if queue is empty
+        """
+
+        read_address, write_address = control_info.queue_info()
+        read_index, = await self.read_memory(read_address, 1, chunk_size=1)
+        write_index, = await self.read_memory(write_address, 1, chunk_size=1)
+
+        if read_index == write_index:
+            return True
+
+        try:
+            await self._read_next_queue_frame(control_info, read_index)
+        except (HardwareError, pylink.errors.JLinkException):
+            logger.debug("Queue poll exception.", exc_info=True)
+        except:
+            logger.exception("Unexpected queue poll exceptioin!")
+
+        read_index = (read_index + 1) % control_info.QUEUE_SIZE
+        await self.write_memory(read_address, [read_index], chunk_size=1)
+
+        return read_index == write_index
+
+    async def _update_watch_counter(self, control_info):
+        counter_address = control_info.counter_info()
+
+        counter, = await self.read_memory(counter_address, 1, chunk_size=1)
+        counter = (counter + 1) % 256
+        await self.write_memory(counter_address, [counter], chunk_size=1)
+
     async def _maintenance_coroutine(self, control_info, step_timeout):
         while self._maintenance_counter > 0:
-            if self._jlink_adapter.opened_interfaces["rpc"]:
+            try:
                 await self._poll_rpc_status(control_info)
+                while not await self._poll_queue_status(control_info):
+                    pass
+                await self._update_watch_counter(control_info)
+            except concurrent.futures.CancelledError:
```

Comment:
This should not be the exception that you receive here.  `concurrent.futures.CancelledError` is thrown only by multithreading code.  Since we're using asyncio tasks, the exception will be `asyncio.CancelledError`.

Reply:
Fixed it.